kvdb-rocksdb/src/lib.rs

```diff
@@ -152,47 +161,80 @@ impl CompactionProfile {
 pub struct DatabaseConfig {
 	/// Max number of open files.
 	pub max_open_files: i32,
-	/// Memory budget (in MiB) used for setting block cache size, write buffer size.
-	pub memory_budget: Option<usize>,
-	/// Compaction profile
+	/// Memory budget (in MiB) used for setting block cache size and
+	/// write buffer size for each column including with the default one.
+	/// If the memory budget of a column is not specified,
+	/// `DB_DEFAULT_COLUMN_MEMORY_BUDGET_MB` is used for that column.
+	pub memory_budget: HashMap<Option<u32>, MiB>,
+	/// Compaction profile.
 	pub compaction: CompactionProfile,
-	/// Set number of columns
+	/// Set number of columns.
 	pub columns: Option<u32>,
+	/// Specify the maximum number of info/debug log files to be kept.
+	pub keep_log_file_num: i32,
 }
 
 impl DatabaseConfig {
 	/// Create new `DatabaseConfig` with default parameters and specified set of columns.
 	/// Note that cache sizes must be explicitly set.
 	pub fn with_columns(columns: Option<u32>) -> Self {
-		let mut config = Self::default();
-		config.columns = columns;
-		config
+		Self { columns, ..Default::default() }
+	}
+
+	/// Returns the total memory budget in bytes.
+	pub fn memory_budget(&self) -> MiB {
+		if self.memory_budget.is_empty() && self.columns.is_none() {
+			return DB_DEFAULT_MEMORY_BUDGET_MB * MB;
+		}
+		(0..=self.columns.unwrap_or(0))
+			.map(|i| self.memory_budget.get(&i.checked_sub(1)).unwrap_or(&DB_DEFAULT_COLUMN_MEMORY_BUDGET_MB) * MB)
+			.sum()
 	}
 
-	pub fn memory_budget(&self) -> usize {
-		self.memory_budget.unwrap_or(DB_DEFAULT_MEMORY_BUDGET_MB) * MB
+	/// Returns the memory budget of the specified column in bytes.
+	fn memory_budget_per_col(&self, col: Option<u32>) -> MiB {
+		self.memory_budget.get(&col).unwrap_or(&DB_DEFAULT_COLUMN_MEMORY_BUDGET_MB) * MB
 	}
 
-	pub fn memory_budget_per_col(&self) -> usize {
-		self.memory_budget() / self.columns.unwrap_or(1) as usize
+	// Get column family configuration with the given block based options.
+	fn column_config(&self, block_opts: &BlockBasedOptions, col: Option<u32>) -> io::Result<Options> {
+		let memory_budget_per_col = self.memory_budget_per_col(col);
+		let mut opts = Options::new();
+
+		opts.set_parsed_options("level_compaction_dynamic_level_bytes=true").map_err(other_io_err)?;
+
+		opts.set_block_based_table_factory(block_opts);
+
+		opts.set_parsed_options(&format!(
+			"block_based_table_factory={{{};{}}}",
+			"cache_index_and_filter_blocks=true", "pin_l0_filter_and_index_blocks_in_cache=true"
+		))
+		.map_err(other_io_err)?;
+
+		opts.optimize_level_style_compaction(memory_budget_per_col as i32);
+		opts.set_target_file_size_base(self.compaction.initial_file_size);
+
+		opts.set_parsed_options("compression_per_level=").map_err(other_io_err)?;
```

Comment:
This was added in https://github.com/paritytech/parity-ethereum/pull/7348/commits/3cf1aac5b9e8fe23c68e588ee1a74304a7fecf5c by @andresilva with the intention of enabling compression on all levels, otherwise optimize_level_style_compaction [will disable](https://github.com/facebook/rocksdb/blob/6ec6a4a9a49e506eff76aebd104d30be6a2d36cc/options/options.cc#L548-L550) compression for the first two levels IIUC. #257 uses https://docs.rs/rocksdb/0.13.0/rocksdb/struct.Options.html#method.set_compression_per_level instead of `set_parsed_option`.